File: web/src/services/roles.js
// Mirrors functions/index.js ROLES exactly and app/.../data/model/Roles.kt —
// these three must always agree, since the string values are also the Firebase
// custom claim values that the Cloud Functions enforce against.

export const ROLES = {
  SUPER_ADMIN: 'SUPER_ADMIN',
  ADMIN: 'ADMIN',
  ENTREPRENEURIAL_MEMBER: 'ENTREPRENEURIAL_MEMBER',
  GENERAL_MEMBER: 'GENERAL_MEMBER',
  NEW_MEMBER: 'NEW_MEMBER',
};

const RANK = {
  SUPER_ADMIN: 4,
  ADMIN: 3,
  ENTREPRENEURIAL_MEMBER: 2,
  GENERAL_MEMBER: 1,
  NEW_MEMBER: 0,
};

export function rankOf(role) {
  return RANK[role] ?? 0;
}

// True for both ADMIN and SUPER_ADMIN — matches the server's requireAdmin().
export function isAdminLevel(role) {
  return role === ROLES.ADMIN || role === ROLES.SUPER_ADMIN;
}

export function isSuperAdmin(role) {
  return role === ROLES.SUPER_ADMIN;
}

const DISPLAY_EN = {
  SUPER_ADMIN: 'Super Admin',
  ADMIN: 'Admin',
  ENTREPRENEURIAL_MEMBER: 'Entrepreneurial Member',
  GENERAL_MEMBER: 'General Member',
  NEW_MEMBER: 'New Member',
};

const DISPLAY_BN = {
  SUPER_ADMIN: 'সুপার অ্যাডমিন',
  ADMIN: 'অ্যাডমিন',
  ENTREPRENEURIAL_MEMBER: 'উদ্যোক্তা সদস্য',
  GENERAL_MEMBER: 'সাধারণ সদস্য',
  NEW_MEMBER: 'নতুন সদস্য',
};

export function displayName(role, lang = 'en') {
  const table = lang === 'bn' ? DISPLAY_BN : DISPLAY_EN;
  return table[role] || role || (lang === 'bn' ? 'সদস্য' : 'Member');
}

File: app/src/main/java/com/example/data/model/Roles.kt
package com.example.data.model

/**
 * The five-tier role hierarchy, mirrored exactly from functions/index.js (ROLES).
 * These string values must match the Firebase custom claim / Firestore `role`
 * field set server-side — never invent a different casing/spelling here, or
 * role checks will silently fail open or closed depending on the bug.
 */
object Roles {
    const val SUPER_ADMIN = "SUPER_ADMIN"
    const val ADMIN = "ADMIN"
    const val ENTREPRENEURIAL_MEMBER = "ENTREPRENEURIAL_MEMBER"
    const val GENERAL_MEMBER = "GENERAL_MEMBER"
    const val NEW_MEMBER = "NEW_MEMBER"

    private val RANK = mapOf(
        SUPER_ADMIN to 4,
        ADMIN to 3,
        ENTREPRENEURIAL_MEMBER to 2,
        GENERAL_MEMBER to 1,
        NEW_MEMBER to 0
    )

    fun rankOf(role: String?): Int = RANK[role] ?: 0

    /** True for both ADMIN and SUPER_ADMIN — matches the server's requireAdmin(). */
    fun isAdminLevel(role: String?): Boolean = role == ADMIN || role == SUPER_ADMIN

    fun isSuperAdmin(role: String?): Boolean = role == SUPER_ADMIN

    fun displayName(role: String?, bengali: Boolean = false): String = when (role) {
        SUPER_ADMIN -> if (bengali) "সুপার অ্যাডমিন" else "Super Admin"
        ADMIN -> if (bengali) "অ্যাডমিন" else "Admin"
        ENTREPRENEURIAL_MEMBER -> if (bengali) "উদ্যোক্তা সদস্য" else "Entrepreneurial Member"
        GENERAL_MEMBER -> if (bengali) "সাধারণ সদস্য" else "General Member"
        NEW_MEMBER -> if (bengali) "নতুন সদস্য" else "New Member"
        else -> role ?: (if (bengali) "সদস্য" else "Member")
    }
}
